### ISSUE TITLE
Amortize document updates ingestion costs

### ISSUE THREAD

**Issue by Kerollmops** [meilisearch/MeiliSearch#88]
The current update ingestion implementation recompute the reverse index.
It is an important CPU cost mostly for adding a small number of documents.

We can't deal with this anymore!
We must be able to acppet any size updates and make them be available to search without the minimum overhead.

We thought about something similar to database caching algorithms: putting most recently used entries in memory providing a rapid access to it. When too much entries are cached, a compaction is performed (i.e. some of the [Sled's future strategy][1]).

So, basically when a user propose some documents update/insertion/deletion we will not put every document attributes updates inside of an SST file, paying the cost of keeping words-docindex ordered in RAM and constructing a new [FST][1].

Inserting words-docindex and document attributes values directly into the key-value store is the solution. Internally the database will extract words of the indexed attributes and update the special "revindex" keyspace, dedicated to reverse index data, storing `DocIndex` associated to words as keys-values, like the current FST.

The advantage of a key-value store is that, like a BTree and an FST, keys are ordered, making `O(log n)` *unions* possible between these datastructures **at search time**.

Obviously, we should find a rule to trigger compaction, it must be based on the latency involved in doing the union, probably related to the number of words/doc-indexes stored as raw keys. A compaction will be an update of the FST by integrating "revindex" word-docindex into the FST.

[1]: https://github.com/spacejam/sled/wiki/sled-architectural-outlook#caching
[2]: https://github.com/BurntSushi/fst

**Comment by alexisvisco** [comment]
You should look at this talk https://youtu.be/s19G6n0UjsM?t=814. 

Database caching is a complex stuff and this person done research at MIT about do smart and well caching. It precomputes and caches query results. [His database](https://github.com/mit-pdos/noria) automatically keeps cached results up-to-date as the underlying data, stored in persistent base tables, change. 

His way of thinking is absolutely insane and I recommend you to look at the video that I have shared.

**Comment by Kerollmops** [comment]
Haha! Thank you the good stuff you and share.
That's one of my favorite talk, I often follow its Youtube channel, he is currently working on implementing a concurrent lock-free HashMap from the Java world, quite fun to look at that, great stuff! He also shared the research paper about Noria that I read, in part.

This issue is a outdated it was refering of the old way I stored the postings list (all the positions where words matched for a given word) and the FST (the set of all the words, the dictionnary) in the same memory area. That was no quite good idea.

The current system stores the FST and the each postings lists in seperate entries in LMDB, one key for each postings list and another for the FST, this way when we even add a little amount of documents we only have to recompute the FST ([that's pretty fast](https://docs.rs/fst/0.3.5/fst/struct.SetBuilder.html#method.extend_stream)) and only the postings lists refering to the words contained in the added documents, [I have made a generic library to do this kind of operations fast on slices](https://github.com/Kerollmops/sdset).

Thank you for your job again, I'm closing this as it is no longer relevant :)
